public/app/plugins/datasource/loki/querybuilder/components/LabelBrowserModal.tsx

```diff
@@ -36,13 +41,42 @@ export const LabelBrowserModal = (props: Props) => {
     onClose();
   };
 
+  const getChooserText = (logLabelsLoaded: boolean, hasLogLabels: boolean) => {
+    if (!logLabelsLoaded) {
+      return 'Loading labels...';
+    }
+    if (!hasLogLabels) {
+      return '(No labels found)';
+    }
+    return 'Label browser';
+  };
+
+  const hasLogLabels = datasource.languageProvider.getLabelKeys().length > 0;
+  const labelBrowserText = getChooserText(labelsLoaded, hasLogLabels);
+
+  if (labelBrowserText === 'Loading labels...') {
+    return (
+      <Modal title="Label browser" isOpen={isOpen} onDismiss={onClose}>
+        <p>Loading labels...</p>
+      </Modal>
+    );
+  }
+
+  if (labelBrowserText === '(No labels found)') {
+    return (
+      <Modal title="Label browser" isOpen={isOpen} onDismiss={onClose}>
+        <p>No labels found.</p>
+      </Modal>
+    );
```

Comment:
Here it seems that you are duplicating logic. Instead of creating `labelBrowserText` by calling `getChooserText`, you could always return modal, just content would change:

- labels are not loaded, return `<p>Loading labels...</p>`
- labels are loaded but empty, return `<p>No labels found.</p>`
- labels are loaded and not empty, return `<LocalStorageValueProvider....`

Reply:
Also to `loading labels`, you could add spinner.